File: 프로그래머스/lv2/68936. 쿼드압축 후 개수 세기/쿼드압축 후 개수 세기.py
one,zero=0,0

def solution(arr):
    answer = []
    
    def check(x,y,size):
        temp=arr[x][y]
        for i in range(x,x+size):
            for j in range(y,y+size):
                if temp!=arr[i][j]:
                    return False;
        return True;
    
    def dfs(x,y,size):
        global one,zero
        if check(x,y,size):
            if arr[x][y]==1:
                one+=1
            else: zero+=1
            return
        
        newSize=size//2
        dfs(x,y,newSize)
        dfs(x,y+newSize,newSize)
        dfs(x+newSize,y,newSize)
        dfs(x+newSize,y+newSize,newSize)
    dfs(0,0,len(arr))
        
    return zero,one

File: 프로그래머스/lv2/68936. 쿼드압축 후 개수 세기/쿼드압축 후 개수 세기.java
class Solution {
    static int one;
    static int zero;
    public int[] solution(int[][] arr) {
        int[] answer = new int[2];
        one=0;
        zero=0;
        quard(arr.length,arr,0,0);
        answer[0]=zero;
        answer[1]=one;
        return answer;
    }
    public void quard(int size,int[][] arr,int x,int y){
        if(check(arr,size,x,y)){
            if(arr[x][y]==0)
                zero++;
            else one++;
            return;
        }
        int newSize=size/2;
        quard(newSize,arr,x,y);
        quard(newSize,arr,x+newSize,y+newSize);
        quard(newSize,arr,x+newSize,y);
        quard(newSize,arr,x,y+newSize);
    }
    public boolean check(int[][] arr,int size,int x,int y){
        int temp=arr[x][y];
        for(int i=x;i<x+size;i++){
            for(int j=y;j<y+size;j++){
                if(temp!=arr[i][j])
                    return false;
            }
        }
        return true;
    }
}